### ISSUE TITLE
GNN Training Infrastructure Missing Forgetting Mitigation Logic

### ISSUE THREAD

**Issue by ruvnet** [ruvnet/RuVector#17]
## Summary

The GNN training infrastructure has working loss functions and SGD steps, but lacks critical forgetting mitigation logic. The `propagate_updates` flag in `OnlineConfig` is a dead flag that is never consumed.

### Implemented (Working)
- `info_nce_loss(anchor, positives, negatives, temperature)` - Loss computation
- `local_contrastive_loss(node, neighbors, non_neighbors, temperature)` - Loss computation  
- `sgd_step(embedding, gradient, learning_rate)` - Single SGD step
- Gradient accumulation in `mmap.rs` - Pure SGD, no momentum

### NOT Implemented (Stubs/Dead Code)

```rust
// training.rs:37-39
pub fn step(&mut self, params: &mut Array2<f32>, grads: &Array2<f32>) -> Result<()> {
    unimplemented!("TODO: Implement optimizer step")  // <-- DEAD
}

// training.rs:58-63
pub fn compute(...) -> Result<f32> {
    unimplemented!("TODO: Implement loss computation")  // <-- DEAD
}

// OnlineConfig.propagate_updates - never read
// OnlineConfig.local_steps - never read
```

## The Forgetting Problem

When query patterns shift:

1. **Time 0**: Queries mostly hit region A → GNN learns A relationships
2. **Time 1**: Traffic shifts to region B → SGD updates favor B  
3. **Time 2**: Region A relationships are overwritten → **catastrophic forgetting**

Current mitigation: **None**

The only implicit protection is graph topology (neighbors always available as positives), but that's just data, not learned weights. Temperature 0.07 with sharp gradients makes forgetting worse.

## Required Implementations

| Component | Status | Priority |
|-----------|--------|----------|
| Loss functions | ✅ Working | - |
| SGD step | ✅ Working | - |
| Gradient accumulation | ✅ Working | - |
| Optimizer (Adam/momentum) | ❌ `unimplemented!()` | High |
| Replay buffer | ❌ Not present | High |
| EWC/Fisher penalty | ❌ Not present | Medium |
| Learning rate scheduling | ❌ Not present | Medium |
| `propagate_updates` consumption | ❌ Dead flag | Low |

### 1. Replay Buffer
```rust
struct ReplayBuffer {
    queries: VecDeque<(Vec<f32>, Vec<usize>)>,  // (query, positive_ids)
    capacity: usize,
}
```

### 2. Elastic Weight Consolidation (EWC)
```rust
struct ElasticWeightConsolidation {
    fisher_diag: Vec<f32>,           // Fisher information diagonal
    anchor_weights: Vec<f32>,         // Weights to regularize toward
    lambda: f32,                      // Regularization strength
}
```

### 3. Momentum/Adam Optimizer
```rust
struct AdamState {
    m: Vec<f32>,  // First moment
    v: Vec<f32>,  // Second moment
    t: usize,     // Timestep
}
```

## Production Risk

Under distribution shift, the system will forget previously learned relationships with no mitigation. The architecture assumes stationary query distributions, which is unrealistic in production environments.

## Acceptance Criteria

- [ ] Implement `Optimizer::step` with Adam or momentum
- [ ] Add replay buffer for experience replay during training
- [ ] Implement EWC penalty or similar continual learning technique
- [ ] Add learning rate scheduling
- [ ] Either implement `propagate_updates` logic or remove the dead flag
- [ ] Add tests for forgetting mitigation under distribution shift

**Comment by ruvnet** [comment]
The forgetting mitigation logic described here has since been implemented:

- **Optimizer** (SGD with momentum + Adam) — fully implemented in `crates/ruvector-gnn/src/training.rs`
- **Loss functions** (MSE, CrossEntropy, BCE, InfoNCE, contrastive) — implemented with numerical stability guards
- **Experience Replay Buffer** — `crates/ruvector-gnn/src/replay.rs` with reservoir sampling for uniform distribution
- **LR Schedulers** — `crates/ruvector-gnn/src/scheduler.rs` with cosine annealing, warmup-linear, reduce-on-plateau

The `OnlineConfig.propagate_updates` and `local_steps` fields exist in the struct but are not yet consumed by an online-update function — that would be the remaining work if graph-topology propagation is desired.
